File: python/Data Structures and Algorithms in Python Book/oop/fibonacciprogression.py
from progression import Progression

class FibonacciProgression(Progression):
    
    def __init__(self, first=0, second=1):
        super().__init__(start=first)
        self._previous = second - first
    
    def _advance(self):
        self._previous, self._current = self._current, self._previous + self._current


if __name__ == "__main__":
    fibonacci_progresssion = FibonacciProgression(first= 1, second= 2)
    fibonacci_progresssion.print_progression(20)

File: java/ood/FibonacciProgression.java
package ood;

public class FibonacciProgression extends Progression {
    protected long previous;

    /** Construct traditional Fibonacci progression starting 0, 1, 1, 2, 3, ... */
    public FibonacciProgression() {
        this(0, 1);
    }

    /**
     * Construct generalized Fibonacci progression, with the given first and second
     * values.
     */
    public FibonacciProgression(long first, long second) {
        super(first);
        // fictious value preceeding the first; as we calculate
        // next = previous + current
        previous = second - first;
    }

    /** Replaces (previous, current) with (current, current + previous). */
    @Override
    protected void advance() {
        long temp = previous;
        previous = current;
        // add the value of prevoius that was before
        // the assignment of previous = current
        current += temp;
    }
}